src/python/zquantum/qaoa/estimators/_cvar_estimator.py

```diff
@@ -101,47 +103,45 @@ def _calculate_expectation_value_for_distribution(
     distribution: BitstringDistribution, operator: IsingOperator, alpha: float
 ) -> float:
     # Calculates expectation value per bitstring
-    expectation_values_per_bitstring = {}
-    for bitstring in distribution.distribution_dict:
-        expected_value = _calculate_expectation_value_of_bitstring(bitstring, operator)
-        expectation_values_per_bitstring[bitstring] = expected_value
+    expectation_values = _calculate_expectation_values(
+        np.array([*distribution.distribution_dict.keys()]), operator
+    )
+
+    # Map expectation values back to original bitstrings
+    expectation_values_dict = {
+        bitstring: expectation_values[i]
+        for i, bitstring in enumerate(distribution.distribution_dict.keys())
+    }
 
     return _sum_expectation_values(
-        expectation_values_per_bitstring, distribution.distribution_dict, alpha
+        expectation_values_dict, distribution.distribution_dict, alpha
     )
 
 
 def _calculate_expectation_value_for_wavefunction(
     wavefunction: Wavefunction, operator: IsingOperator, alpha: float
 ) -> float:
-    expectation_values_per_bitstring = {}
-    probability_per_bitstring = {}
-
     n_qubits = wavefunction.amplitudes.shape[0].bit_length() - 1
 
-    for decimal_bitstring in range(2 ** n_qubits):
-        # `decimal_bitstring` is the bitstring converted to decimal.
-
-        # Convert decimal bitstring into bitstring
-        bitstring = "".join([str(int) for int in dec2bin(decimal_bitstring, n_qubits)])
-
-        # Calculate expectation values for each bitstring.
-        expected_value = _calculate_expectation_value_of_bitstring(bitstring, operator)
-        expectation_values_per_bitstring[bitstring] = expected_value
+    # Calculate expectation values for each bitstring.
+    expectation_values = _calculate_expectation_values(
+        np.array([dec2bin(n, n_qubits) for n in range(2 ** n_qubits)]),
```

Comment:
[request] Correct me if I'm wrong, but it looks to me that here you're calculating expectation values for all possible bitstrings, while it should be calculated only for the expectation values with non-zero elements in the wavefunction (or maybe below some adjustable epsilon). This is just brute-forcing all possible solutions.

Reply:
Ah it was like this in the old implementation (it [looped over](https://github.com/zapatacomputing/z-quantum-qaoa/blob/main/src/python/zquantum/qaoa/estimators/_cvar_estimator.py#L126) every bitstring & calculated each expectation value separately) so I didn't realize the problem.

Though idrk whether changing this is worth it because the number of bitstrings that have near-zero amplitudes in the wavefunction is generally pretty low (never more than 10% when eps=1e-5). & it doesn't save any time when I tested it, because the overhead from finding the bitstrings with nonzero amplitudes offsets the time savings from a smaller bitstring vector. (or maybe  it's just my way of doing it isn't very efficient)

It also introduces a bit of imprecision (anywhere from 0.01% - 2.5% depending on alpha) (maybe this is because 1e-5 cutoff is kind of high, but if I lower the epsilon there is virtually no bitstring whose amplitude falls below it)

Reply:
[Here's](https://github.com/zapatacomputing/z-quantum-qaoa/pull/88/commits/4796430f7d6b0380e17157b36a49a23ef4f60b93) the changes anyway if you want to take a look

Reply:
It might not be worth it for a small and/or not well-trained QAOA.
But for bigger number of qubits (for 17 it's 2^17=131k hamiltonian evaluations, which might be significant).

One more thing to change: please create a global constant, e.g. `PROBABILITY_CUTOFF=1e-5`. Also, change its value to `1e-8`  – I think it should be enough. This will make it clearer and easier to actually make it parametrizable if we need to do so in the future. Thanks @alexjuda for the suggestion!

Reply:
Done!